### ISSUE TITLE
MIXED_DML_OPERATION Errors Occurring in Customer Unit Tests

### ISSUE THREAD

**Issue by jesslopez-sf** [SalesforceFoundation/NPSP#6551]
### Issue
Some customer unit tests are failing with MIXED_DML_OPERATION errors because our Trigger Handler records are erroneously being inserted within the same context as a setup object (for example, User).

```
EXCEPTION_THROWN|[15]|System.DmlException: Insert failed. 
First exception on row 0; first error: CANNOT_INSERT_UPDATE_ACTIVATE_ENTITY, 
npsp.TDTM_User: execution of BeforeInsert

caused by: System.DmlException: Insert failed. First exception on row 0; 
first error: MIXED_DML_OPERATION, DML operation on setup object is not permitted 
after you have updated a non-setup object 
(or vice versa): npsp__Trigger_Handler__c, original object: User: []

Class.npsp.TDTM_TriggerHandler.run: line 92, column 1
Class.npsp.TDTM_Config_API.run: line 92, column 1
Trigger.npsp.TDTM_User: line 34, column 1: []
```

### Root Cause
When the TDTM_User trigger calls out to TDTM_Config_API’s run() method which calls out to TDTM_TriggerHandler’s run() method, we check to see if the describeObj passed into the Trigger/TDTM is equal to Schema.SObjectType.User. 
https://github.com/SalesforceFoundation/NPSP/blob/a710e68c36a786302c0103c8dad7a1f306447894/src/classes/TDTM_TriggerHandler.cls#L84-L86
If so, we return out of the method; if not, we insert the default Trigger Handlers because we are running in a test context. We explicitly do not insert Trigger Handler records if a User is inserted to avoid Mixed DML errors. However, in some customer orgs, the comparison of describeObj to Schema.SObjectType.User is returning false and we are inserting the Trigger Handlers. 

Sample Debug Log Lines:
```
10:16:45.105 (2128599156)|METHOD_EXIT|[82]|01p1N000009UvKK|npsp.TDTM_ObjectDataGateway.isEmpty()
10:16:45.105 (2128606972)|STATEMENT_EXECUTE|[82]
10:16:45.105 (2128835915)|HEAP_ALLOCATE|[84]|Bytes:4
10:16:45.105 (2128997899)|STATEMENT_EXECUTE|[84]
10:16:45.105 (2129000462)|STATEMENT_EXECUTE|[87]
```
(It should hit Line 85 here because the values _should_ be equal, but it doesn't.

`describeObj` is a `Schema.DescribeSObjectResult` and the documentation states: (https://developer.salesforce.com/docs/atlas.en-us.apexref.meta/apexref/apex_methods_system_sobject_describe.htm)

> Instances of describe results on the same DescribeSObjectResult aren’t guaranteed to be equal because the state and behavior of a describe object is determined by various factors including the API version used. To compare describe results, call the getSObjectType() method on the SObject describe results and use the equality operator (==) to compare the SObjectType (https://developer.salesforce.com/docs/atlas.en-us.apexref.meta/apexref/apex_class_Schema_SObjectType.htm#topic-title) values.

I’m requesting that we change the comparison to instead use the following syntax:
```
if (describeObj.getsObjectType() == User.sObjectType) {
   return;
}
```

### Steps to Repeat
In a sandbox org (I’m unable to replicate in a scratch org with both managed and unmanaged NPSP installed):

* Go to Setup → Roles and create a Role named “Test Role”.
* Create an Apex Class named UserRoleTest with the following code:
```
@isTest 
private class UserRoleTest {
    
     @isTest 
     private static void createUserWithRoleTest() {
         Id roleId = [SELECT Id From UserRole  WHERE Name = 'Test Role'].Id;
         Id adminId = [Select Id From Profile Where Name = 'System Administrator' Limit 1].Id;
        
          User u = new User(FirstName = 'Test', LastName='User', Username = 'sfdotest@example.com',
                             UserRoleId = RoleId, ProfileId = adminId, Email = 'sfdotest@example.com', Alias = 'test1', EmailEncodingKey = 'UTF-8',LanguageLocaleKey = 'en_US', LocaleSidKey = 'en_US',TimeZoneSidKey = 'America/Los_Angeles' );
         
         Test.startTest();
         insert u;
         Test.stopTest();
         
       
         List<npsp__Trigger_Handler__c> th = [SELECT Id FROM npsp__Trigger_Handler__c];
         System.assertEquals(0, th.size());
     }
    
}
```
* Run the test and it should fail with a MIXED_DML_OPERATION error.

### Potential Workaround
Until this is resolved in NPSP, there is a potential workaround depending upon how customers have set up their unit tests. If the Users are not being inserted in a testSetup method, you can upsert our default Trigger Handler records within the testSetup method so that it isn't part of the same transaction as the User insert.

For example, add something like the following:
```
private static void createInitialData(){
        List<npsp__Trigger_Handler__c> listHandlers = npsp.TDTM_Config_API.getCachedRecords();
		upsert listHandlers;
}
```

**Comment by jesslopez-sf** [comment]
W-9625866

**Comment by cdonnelly-cms** [comment]
I'm experiencing this issue.

**Comment by jesslopez-sf** [comment]
Just added a potential workaround of upserting the NPSP default Trigger Handlers in a testSetup method. It won't work for everyone (depending on their unit test setup/configuration), but can provide a temporary fix until we resolve in the product.

**Comment by cdonnelly-cms** [comment]
In my case, the error is occurring at the end of the testSetup method which has already inserted a user record (the fbc_SearchUtilsTest unit test from the Flow Base Components pack https://github.com/alexed1/LightningFlowComponents/blob/master/flow_process_components/FlowBaseComponents/force-app/main/default/classes/fbc_SearchUtilsTest.cls).

**Comment by cdonnelly-cms** [comment]
Any news on when a fix might be pushed into production orgs? This issue is preventing me from deploying any code changes to production.

**Comment by AndersonKatharineTNC** [comment]
This is happening in my sandbox when I try to create a user record, but fortunately is not causing issues yet in production.

**Comment by schlossmj** [comment]
The fix for this issue will go out in NPSP 3.206.

**Comment by jesslopez-sf** [comment]
@cdonnelly-cms and @AndersonKatharineTNC  - just wanted to let you know that this fix should be hitting your sandboxes within a few hours or so. It definitely resolved the error in my sandbox org where I replicated. If the issue still occurs in yours, please let us know asap. Thank you so much!
